### PR TITLE
Optional checkpointing

### PR DESCRIPTION
**Pull request by kushti**
This PR is generalizing optional checkpointing introduced after the hard-fork. In this PR, it is possible to set arbitrary checkpoint. Scripts in blocks before the checkpoint will not be validated, correctness of UTXO set is checked anyway. Thus bootstrapping can be significantly improved with no security issues.

### CODE REVIEW

File: src/main/resources/mainnet.conf [hunk 1]
```diff
@@ -30,7 +30,10 @@ ergo {
     # Skip validation of transactions in the mainnet before block 417,792 (in v1 blocks).
```
Comment (greenhat) [inline src/main/resources/mainnet.conf:30]:
```suggestion
    # Skip validation of transactions in the mainnet before the given block (height)
```

File: src/main/resources/mainnet.conf [hunk 2]
```diff
@@ -30,7 +30,10 @@ ergo {
...
     # Block 417,792 is checkpointed by the protocol (so its UTXO set as well).
```
Comment (greenhat) [inline src/main/resources/mainnet.conf:31]:
```suggestion
    # The given block is checkpointed by the protocol (so its UTXO set as well).
```

File: src/main/scala/org/ergoplatform/settings/ValidationRules.scala [hunk 3]
```diff
@@ -140,7 +140,7 @@ object ValidationRules {
     hdrVotesUnknown -> RuleStatus(s => fatal(s"First header of an epoch should not contain a vote for unknown parameter. $s"),
       Seq(classOf[Header]),
       mayBeDisabled = true),
-    hdrCheckpointV2 -> RuleStatus(s => fatal(s"First version 2 header on the mainnet on height 417,729 should have " +
+    hdrCheckpoint -> RuleStatus(s => fatal(s"First version 2 header on the mainnet on height 417,729 should have " +
```
Comment (greenhat) [inline src/main/scala/org/ergoplatform/settings/ValidationRules.scala:143]: Should the hard-coded height be changed to the one set in the config file?

**Review by greenhat — CHANGES_REQUESTED** [review]

File: src/main/resources/application.conf [hunk 4]
```diff
@@ -82,10 +82,7 @@ ergo {
     # default value is 100 blocks ~= 200 minutes
     headerChainDiff = 100
 
-    # Skip validation of transactions in the mainnet before block 417,792 (in v1 blocks).
-    # Block 417,792 is checkpointed by the protocol (so its UTXO set as well).
-    # The node still applying transactions to UTXO set and so checks UTXO set digests for each block.
-    skipV1TransactionsValidation = false
+    checkpoint = null
```
Comment (aslesarenko) [inline src/main/resources/application.conf:85]: Please add a description.

File: src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/HeadersProcessor.scala [hunk 5]
```diff
@@ -331,19 +331,21 @@ trait HeadersProcessor extends ToDownloadProcessor with ScorexLogging with Score
...
     /**
       * Helper method to check v2 checkpoint (first v2 block)
       */
-    private def checkpointV2Condition(header: Header): Boolean = {
-      if (header.height == 417792 && settings.networkType.isMainNet) {
-        header.id == "0ba60a7db44877aade553beb05200f7d67b586945418d733e455840d283e0508"
-      } else {
-        true
-      }
+    private def checkpointCondition(header: Header): Boolean = {
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/HeadersProcessor.scala:341]: Please add comment that the method returns true only when Id at the checkpoint height is not as expected.

File: src/main/resources/mainnet.conf (hunk 2 above)
Comment (kushti) [reply src/main/resources/mainnet.conf]: Changed comment in another way

File: src/main/resources/mainnet.conf (hunk 1 above)
Comment (kushti) [reply src/main/resources/mainnet.conf]: Changed comment in another way

File: src/main/scala/org/ergoplatform/settings/ValidationRules.scala (hunk 3 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/settings/ValidationRules.scala]: Changed comment

File: src/main/resources/application.conf (hunk 4 above)
Comment (kushti) [reply src/main/resources/application.conf:97]: Done

File: src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/HeadersProcessor.scala (hunk 5 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/HeadersProcessor.scala:343]: done

**Review by greenhat — APPROVED** [review]

**Review by aslesarenko — APPROVED** [review]

**Pull request merged**